launch/launch.go

```diff
@@ -1,21 +1,113 @@
 package launch
 
 import (
+	"encoding/json"
+	"fmt"
 	"path"
 	"path/filepath"
 	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/buildpacks/lifecycle/internal/encoding"
+
+	"github.com/BurntSushi/toml"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
 )
 
 type Process struct {
+	Type             string         `toml:"type" json:"type"`
+	Command          []string       `toml:"-" json:"-"` // ignored
+	RawCommandValue  toml.Primitive `toml:"command" json:"command"`
+	Args             []string       `toml:"args" json:"args"`
+	Direct           bool           `toml:"direct" json:"direct"`
+	Default          bool           `toml:"default,omitempty" json:"default,omitempty"`
+	BuildpackID      string         `toml:"buildpack-id" json:"buildpackID"`
+	WorkingDirectory string         `toml:"working-dir,omitempty" json:"working-dir,omitempty"`
+}
+
+// processSerializer is used to encode a process to toml.
+type processSerializer struct {
 	Type             string   `toml:"type" json:"type"`
-	Command          string   `toml:"command" json:"command"`
+	Command          string   `toml:"command" json:"command"` // command is string
 	Args             []string `toml:"args" json:"args"`
 	Direct           bool     `toml:"direct" json:"direct"`
 	Default          bool     `toml:"default,omitempty" json:"default,omitempty"`
 	BuildpackID      string   `toml:"buildpack-id" json:"buildpackID"`
 	WorkingDirectory string   `toml:"working-dir,omitempty" json:"working-dir,omitempty"`
 }
 
+// MarshalText implements the toml TextMarshaler interface to allow us more control when writing a Process to a toml file.
+func (p Process) MarshalText() ([]byte, error) {
+	return encoding.MarshalTOML(processSerializer{
+		Type:             p.Type,
+		Command:          p.Command[0],
+		Args:             append(p.Command[1:], p.Args[0:]...),
+		Direct:           p.Direct,
+		Default:          p.Default,
+		BuildpackID:      p.BuildpackID,
+		WorkingDirectory: p.WorkingDirectory,
+	})
+}
+
+// MarhsalJSON implements the json Marshaler interface to allow us more control when writing a Process to a json file.
+func (p Process) MarshalJSON() ([]byte, error) {
+	return json.Marshal(processSerializer{
+		Type:             p.Type,
+		Command:          p.Command[0],
+		Args:             append(p.Command[1:], p.Args[0:]...),
+		Direct:           p.Direct,
+		Default:          p.Default,
+		BuildpackID:      p.BuildpackID,
+		WorkingDirectory: p.WorkingDirectory,
+	})
+}
+
+// UnmarshalTOML implements the toml Unmarshaler interface to allow us more control when reading a Process from toml.
+func (p *Process) UnmarshalTOML(data interface{}) error {
+	var tomlString string
+	switch v := data.(type) {
+	case string:
+		tomlString = v
+	case map[string]interface{}:
+		// turn back into a string
+		bytes, _ := encoding.MarshalTOML(v)
+		tomlString = string(bytes)
```

Comment:
Is this block necessary? I thought `pProcess` allows us to avoid double looping through this function. Or at least, when I comment it out, everything still passes...

Reply:
Commenting out the `case map[string]interface{}` path leads to an error.
```
=== RUN   TestDecodeMetadataTOML/DecodeMetadataTOML/decoding_metadata.toml/string_commands/decodes_string_commands_into_command_array

    /Users/jesse.brown/dev/buildpacks/lifecycle/launch/decode_test.go:83: Expected nil: could not cast data to string
```

Reply:
Hmm... that's weird. I guess I don't understand what's going on here :(

Reply:
I think this changes depending on the path it took to get here. If you were to serialize _just_ a `launch.Process` back and forth it would always be a string. If you have a collection of these in a parent struct it ends up with a string for the parent, then children get a map of the that section to process.

Reply:
Ahh that makes sense... thanks for the explanation. Could we add a comment for posterity?